components/raftstore-v2/src/operation/command/admin/merge/prepare.rs

```diff
@@ -276,6 +298,141 @@ impl<EK: KvEngine, ER: RaftEngine> Peer<EK, ER> {
         })
     }
 
+    fn check_trim_status(
+        &mut self,
+        ctx: PreProposeContext,
+        req: &mut RaftCmdRequest,
+    ) -> Result<PreProposeContext> {
+        if self.storage().has_dirty_data() {
+            return Err(box_err!(
+                "source peer {} not trimmed, skip merging.",
+                self.peer_id()
+            ));
+        }
+        let target = req.get_admin_request().get_prepare_merge().get_target();
+        let mut pending_source = HashSet::default();
+        let peers = self.region().get_peers().to_owned();
+        for p in peers {
+            if p.get_id() == self.peer_id() {
+                continue;
+            }
+            let mut msg = RaftMessage::default();
+            msg.set_region_id(self.region_id());
+            msg.set_from_peer(self.peer().clone());
+            msg.set_to_peer(p.clone());
+            msg.set_region_epoch(self.region().get_region_epoch().clone());
+            msg.mut_extra_msg()
+                .set_type(ExtraMessageType::MsgAvailabilityRequest);
+            msg.mut_extra_msg()
+                .mut_availability_context()
+                .set_from_region_id(self.region_id());
+            self.add_message(msg);
+            pending_source.insert(p.get_id());
+        }
+        let mut pending_target = HashSet::default();
+        for p in target.get_peers() {
+            let mut msg = RaftMessage::default();
+            msg.set_region_id(target.get_id());
+            msg.set_from_peer(self.peer().clone());
+            msg.set_to_peer(p.clone());
+            msg.set_region_epoch(target.get_region_epoch().clone());
+            msg.mut_extra_msg()
+                .set_type(ExtraMessageType::MsgAvailabilityRequest);
+            msg.mut_extra_msg()
+                .mut_availability_context()
+                .set_from_region_id(self.region_id());
+            self.add_message(msg);
+            pending_target.insert(p.get_id());
+        }
+        self.merge_context_mut().prepare_status = Some(PrepareStatus::WaitForTrimStatus {
+            start_time: Instant::now_coarse(),
+            pending_source,
+            pending_target,
+            ctx,
+            req: Some(mem::take(req)),
+        });
+        Err(Error::PendingPrepareMerge)
```

Comment:
Can it be Enum instead of Err? To me this is the expected behavior and should not be an Err. 
Also it seems it always return the Err, then "self.check_trim_status(r, &mut req)?" will always yield.

Reply:
I agree it's a bit twisty. But it's very concise, and we do need an error to respond to the proposal sender.

> Also it seems it always returns the Err

Yes, we always start the check in this function. If the check has passed, we would have called `already_checked_trim_status` and no need to call `checked_trim_status`. I'll add an assertion here to explain this.

Reply:
> I agree it's a bit twisty. But it's very concise, and we do need an error to respond to the proposal sender.
> 
> > Also it seems it always returns the Err
> 
> Yes, we always start the check in this function. If the check has passed, we would have called `already_checked_trim_status` and no need to call `checked_trim_status`. I'll add an assertion here to explain this.

I'm still confused. 
         ```
   self.check_trim_status(store_ctx, &mut req)?;
            let r = self.check_logs_before_prepare_merge(store_ctx)?;
            self.check_pessimistic_locks(r, &mut req)?
```
If self.check_trim_status(store_ctx, &mut req)?; always return Err, then the code below it will not execute for sure. It's the dead code then.